GRW-392 / Feature/ Add insurance selector component

mehdielket opened the pull request:
## What?

Add a new `<InsuranceSelector>` component to be able to toggle different bundles directly on the offer page.

Here is an example of how to use this component

```React
<InsuranceSelector
  variants={quoteBundle.possibleVariations}
  selectedQuoteBundle={selectedQuoteBundle}
  onChange={(bundle) => setSelectedQuoteBundle(bundle)}
/>
```

* This will consume the `possibleVariations` that part of `QuoteBundle` query.
* PR for update the the `QuoteBundle` in giraffe is [here](https://github.com/HedvigInsurance/giraffe/pull/1333)
* Integrating this component into the offer page will be done as part of [GRW-391]

## Why?

This is to allow members to get a smaller or larger bundle depending on their needs without going through the whole onboarding again.

_Referenced ticket(s): [GRW-392]_

## Demo
![Screenshot 2021-09-23 at 13 24 36](https://user-images.githubusercontent.com/89857278/134498933-c0a07953-92a2-43ca-ae46-5937dbd36724.png)
![Screenshot 2021-09-23 at 13 24 55](https://user-images.githubusercontent.com/89857278/134498936-14328c3f-0e28-40bf-964e-5854adc19aac.png)


[GRW-391]: https://hedvig.atlassian.net/browse/GRW-391
[GRW-392]: https://hedvig.atlassian.net/browse/GRW-392

mehdielket committed:
GRW-392 Add insurance selector component

robinandeer commented on src/client/pages/OfferNew/InsuranceSelector/Selector.stories.tsx:
```diff
@@ -0,0 +1,40 @@
+  },
+}
+
+export const Default = () => {
+  return <Selector insurances={mockInsurances} onChange={(id) => alert(id)} />
+}
```
With these changes you can try it our better:

```suggestion
export const Default = () => {
  return <Selector insurances={mockInsurances} onChange={(id) => alert(id)} />
}export const Default = () => {
  const [insurances, setInsurances] = React.useState(mockInsurances)

  return (
    <Selector
      id="select-bundle"
      insurances={insurances}
      onChange={(id) =>
        setInsurances((prev) =>
          prev.map((item) => ({ ...item, selected: item.id === id })),
        )
      }
    />
  )
}
```

mehdielket committed:
GRW-392 Improve storybook story

Co-Authored-By: Robin Andeer <robin.andeer@gmail.com>

Commit:
Improve a11y/keyboard navigation support

Implement focus + hover states for Card component.
Use CSS grid for layout of toggles.
Use key in favor of keyCode (deprecated).

robinandeer commented:
Here are the relevant radiogroup docs:
https://www.w3.org/TR/2017/WD-wai-aria-practices-1.1-20170628/examples/radio/radio-1/radio-1.html

robinandeer commented:
I will rebase on master after #604 is merged and fix any conflicts -- but pls have a look and review the approach @gustaveen if you have time!

Commit:
Support showing discounts

Include currency in price props.
Fix issue with responsiveness.

Commit:
Use QuoteBundleVariant type

Update GraphQL types.
Get full price from bundle quote.

robinandeer commented:
Updated screenshots with hover state and gross price/discount support.

![Screenshot 2021-09-28 at 16 13 15](https://user-images.githubusercontent.com/1220232/135104793-e82f650d-3445-4f97-8ab9-4a2ef15d2ed3.png)

Commit:
Rename fullPrice => grossPrice

Commit:
Rename variables

gustaveen commented on src/client/pages/OfferNew/InsuranceSelector/Selector.tsx:
```diff
@@ -0,0 +1,105 @@
+  cursor: pointer;
+`
+
+interface Props {
```
Should we be consistent with using `type`? :)

robinandeer replied:
Sure, I do agree that consistency is valuable!

gustaveen commented on src/client/pages/OfferNew/InsuranceSelector/Selector.stories.tsx:
```diff
@@ -0,0 +1,48 @@
+  return (
+    <Selector
+      insurances={insurances}
+      onChange={(id) =>
+        setInsurances((prev) =>
+          prev.map((item) => ({ ...item, selected: item.id === id })),
+        )
```
Feel free to omit, but I like to breaking inline functions out in a `onChangeHandler` function :)

robinandeer replied:
Will do!

gustaveen reviewed (commented):
Looks great guys! 🙌

robinandeer commented on src/client/pages/OfferNew/InsuranceSelector/Card.tsx:
```diff
@@ -0,0 +1,110 @@
+      <NameAndPriceWrapper>
+        <Name>{name}</Name>
+        <PriceWrapper>
+          {fullPrice ? <FullPrice>{fullPrice}</FullPrice> : null}
```
I like to be explicit here since `undefined` is not a valid React child (right?) but `null` is. What do you think?

Commit:
Update to default GraphQL types

Use type for props definition.
Extract onChange handler.

robinandeer committed:
Use `IconRoot` for SVG components

Co-authored-by: Gustav Eén <gustav.een@hedvig.com>

robinandeer commented:
@mehdielket: any additional changes you believe we should make? I updated the types in `InsuranceSelector.index` module to use the default GraphQL types.

Commit:
Use logical AND operator in JSX

Co-Authored-By: Gustav Eén <gustaveen@gmail.com>

mehdielket commented:
> @mehdielket: any additional changes you believe we should make? I updated the types in `InsuranceSelector.index` module to use the default GraphQL types.

This looks great to me! I've tested it with some test code on the offer page it seems to work great. Well done!

gustaveen commented on src/client/pages/OfferNew/InsuranceSelector/Selector.tsx:
```diff
@@ -0,0 +1,105 @@
+        <CardWrapper
+          id={id}
+          key={id}
+          tabIndex={selected ? 0 : -1}
```
What's the logic for tabindex right here? I'm guessing since we set selected manually we remove it from the navigation flow? @robinandeer @mehdielket

robinandeer replied:
The idea is to capture focus on the selected item when you tab to the radio group.

Then you use the arrow keys to change between options. Tabbing again will move focus away from the radio group.

I implemented according to: https://www.w3.org/TR/2017/WD-wai-aria-practices-1.1-20170628/examples/radio/radio-1/radio-1.html

gustaveen commented on src/client/pages/OfferNew/InsuranceSelector/Selector.tsx:
```diff
@@ -0,0 +1,105 @@
+      case 'ArrowDown':
+        event.preventDefault()
+        onChange(
+          currentFocusIndex <= insurances.length - 1 - 1
```
To me this is easier to understand :) 
```suggestion
          currentFocusIndex === insurances.length - 1
            ? insurances[0].id
            : insurances[currentFocusIndex + 1].id,
```

robinandeer replied:
Yeah sure, that makes a lot of sense

gustaveen commented:
Just some minor comments, otherwise I think it's good to go! 🙌

mehdielket committed:
Merge branch 'master' into GRW-392-add-insurance-selector-component

mehdielket committed:
GRW-392 simplify logic for a11y keyboard navigation

Co-authored-by: Gustav Eén <gustav.een@hedvig.com>

gustaveen reviewed (approved):
🎉
